### PR TITLE
Make CallApiAt::call_api_at work at light client

### PR DESCRIPTION
**Pull request by svyatonik**
Overview:
- since `CallApiAt::call_api_at` on full node first initializes execution environment with `initialise_block` call, extracted this into `CallExecutor::contextual_call` method:
  - on full client left it as-is;
  - on light client it checks if we're trying to call method on 'clear state' (that matches state in database) and redirects call to the full client. Execution proof generation also initializes execution environment with `initialise_block`. 'Total proof' includes proof for both calls;
- altered proof-generation functions from `state_machine` for usage in 'total proof' generation context;
- in future this ^^^ modification could probably be used (if required) to chain multiple remote calls OR for remote calls that are executed on 'non-clear state'.

cc @amaurymartiny : this PR makes `curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "state_getMetadata"}' http://localhost:9933` work.

### CODE REVIEW

File: core/client/src/call_executor.rs [hunk 1]
```diff
@@ -139,6 +175,30 @@ where
...
+		at: &BlockId<Block>,
+		method: &str,
+		call_data: &[u8],
+		changes: &mut OverlayedChanges,
+		initialised_block: &mut Option<BlockId<Block>>,
+		prepare_environment_block: PB,
+		manager: ExecutionManager<EM>,
+	) -> Result<Vec<u8>, error::Error> where ExecutionManager<EM>: Clone {
+		let state = self.backend.state_at(*at)?;
+		//TODO: Find a better way to prevent double block initialization
```
Comment (svyatonik) [inline core/client/src/call_executor.rs]: This isn't a new TODO (moved from `client.rs`). Could add a tracking issue, if required, but probably better for TODO author to do this.

File: core/client/src/light/call_executor.rs [hunk 2]
```diff
@@ -134,15 +194,30 @@ pub fn check_execution_proof<Header, E, H>(
...
+		Default::default(),
+		request.header.hash(),
+		Default::default(),
+	);
+	execution_proof_check_on_trie_backend::<H, _>(
+		&trie_backend,
+		&mut changes,
+		executor,
+		"initialise_block",
+		&next_block.encode())?;
```
Comment (bkchr) [inline core/client/src/light/call_executor.rs]:
```suggestion
		&next_block.encode()
	)?;
```

File: core/client/src/light/call_executor.rs [hunk 3]
```diff
@@ -123,7 +142,48 @@ where
...
+	let trie_state = state.try_into_trie_backend()
+		.ok_or_else(|| Box::new(state_machine::ExecutionError::UnableToGenerateProof) as Box<state_machine::Error>)?;
+
+	// prepare execution environment + record preparation proof
+	let mut changes = Default::default();
+	let (_, init_proof) = executor.prove_at_trie_state(
+		&trie_state,
+		&mut changes,
+		"initialise_block",
+		&header.encode())?;
```
Comment (bkchr) [inline core/client/src/light/call_executor.rs]:
```suggestion
		&header.encode()
	)?;
```

**Comment by bkchr** [comment]
@svyatonik sorry, please update the wasm stuff once again and I merge it immediately.

**Comment by svyatonik** [comment]
My bad - added these by mistake. Removed + remerged. Let's wait until CI is passed at least :)

**Pull request merged**